uikit/viewmodels/src/main/java/com/sygic/maps/uikit/viewmodels/navigation/preview/RoutePreviewControlsViewModel.kt

```diff
@@ -0,0 +1,83 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package com.sygic.maps.uikit.viewmodels.navigation.preview
+
+import androidx.lifecycle.*
+import com.sygic.maps.tools.annotations.AutoFactory
+import com.sygic.maps.uikit.viewmodels.common.navigation.preview.RouteDemonstrationManager
+import com.sygic.maps.uikit.viewmodels.common.navigation.preview.state.DemonstrationState
+import com.sygic.maps.uikit.views.navigation.preview.RoutePreviewControls
+import com.sygic.maps.uikit.views.navigation.preview.state.PlayPauseButtonState
+
+/**
+ * A [RoutePreviewControlsViewModel] is a basic ViewModel implementation for the [RoutePreviewControls] class. It listens
+ * to the [RouteDemonstrationManager.demonstrationState] and set appropriate state to the [RoutePreviewControls] view.
+ * It also listens to the [RoutePreviewControls] user interaction callbacks and control the [RouteDemonstrationManager].
+ */
+@AutoFactory
+@Suppress("unused", "MemberVisibilityCanBePrivate")
+open class RoutePreviewControlsViewModel internal constructor(
+    private val routeDemonstrationManager: RouteDemonstrationManager
+) : ViewModel(), RoutePreviewControls.OnPlayPauseStateChangedListener {
+
+    val playPauseButtonState: MutableLiveData<PlayPauseButtonState> = MutableLiveData(PlayPauseButtonState.PLAY)
+
+    private val demonstrationStateObserver = Observer<DemonstrationState> { state ->
+        playPauseButtonState.value =
+            if (state == DemonstrationState.ACTIVE) PlayPauseButtonState.PAUSE else PlayPauseButtonState.PLAY
+    }
+
+    init {
+        routeDemonstrationManager.demonstrationState.observeForever(demonstrationStateObserver)
+    }
+
+    override fun onPlayPauseButtonStateChanged(state: PlayPauseButtonState) {
+        when (state) {
+            PlayPauseButtonState.PLAY -> routeDemonstrationManager.pause()
+            PlayPauseButtonState.PAUSE -> {
+                if (routeDemonstrationManager.demonstrationState.value == DemonstrationState.PAUSED) {
+                    routeDemonstrationManager.unPause()
+                } else {
+                    routeDemonstrationManager.restart()
+                }
+            }
+        }
+    }
+
+    open fun onSpeedButtonClick() {
+        routeDemonstrationManager.speedMultiplier.value = routeDemonstrationManager.speedMultiplier.value!! * 2 % 15
+    }
+
+    open fun onStopButtonClick() {
+        routeDemonstrationManager.stop()
+    }
+
+    override fun onCleared() {
+        super.onCleared()
+
+        routeDemonstrationManager.demonstrationState.removeObserver(demonstrationStateObserver)
+        routeDemonstrationManager.destroy()
```

Comment:
tu sa robi `destroy`, ale start s route infom nevidim. nemalo by to byt pokope?

Reply:
respektive toto je len view VM a to iba ovlada demonstraciu, ale nespusta ju. Toto `destroy` by tu nemalo byt

Reply:
oki, pravdu máš.